File: src/nineDOF_Main.py
#nineDOF_Main.py
import numpy as np
from nineDOF_Plant import plant
from nineDOF_Control import simpleHeadingController, testController, make_control_function
from nineDOF_Parameters import systemParameters, atmosphereParameters
from nineDOF_Visualization import visualizeData


#Create parameters from input file
params = systemParameters()
atm = atmosphereParameters()
    
#Create simulator
sim = plant(params, atm)
    
#Initial state: hovering at origin with small perturbation
state0 = np.zeros(18)
state0[0:3] = [0.0, 0.0, -100.0]  # Position: 100m altitude
state0[3:6] = [0.0, 0.1, 0.0]     # Parafoil angles: small pitch
state0[6:9] = [0.0, 0.1, 0.0]     # Cradle angles: small pitch
state0[9:12] = [10.0, 0.0, -0.5]  # Forward velocity
    
#Run simulation
print("Running parafoil-payload simulation...")
t_final = 50.0  # seconds
dt = 0.1        # seconds
targetLandingPoint = (1000.0, 1000.0) #X and Y of the targeted Landing Point
    
#Creating Control Function
control = make_control_function(simpleHeadingController(targetLandingPoint))

times, states = sim.run_simulation(state0, t_final, dt, control) #0.94 is maximum control defelction 
    
print(f"Simulation complete: {len(times)} time steps")
print(f"Final position: x={states[-1,0]:.2f}, y={states[-1,1]:.2f}, z={states[-1,2]:.2f}")
print(f"Final velocity: u={states[-1,9]:.2f}, v={states[-1,10]:.2f}, w={states[-1,11]:.2f}")
    
#Simple visualization
visualizeData.plot_trajectory(states)

File: src/cpp/nineDOF_Main.cpp
#include "nineDOF_Plant.h"
#include "nineDOF_Parameters.h"
#include <iostream>
#include <fstream>
#include <iomanip>

// Example control function
std::array<double, 3> testController(double t, const State& state) {
    // Returns [deltaL, deltaR, incidence]
    return {0.0, 0.0, 0.0};  // No control
}

// Simple visualization - save to CSV
void saveTrajectory(const std::vector<double>& times, 
                   const std::vector<State>& states,
                   const std::string& filename = "trajectory.csv") {
    std::ofstream file(filename);
    
    file << std::fixed << std::setprecision(6);
    file << "time,x,y,z,phi_p,theta_p,psi_p,phi_c,theta_c,psi_c,u,v,w,p_p,q_p,r_p,p_c,q_c,r_c\n";
    
    for (size_t i = 0; i < times.size(); i++) {
        file << times[i];
        for (int j = 0; j < 18; j++) {
            file << "," << states[i][j];
        }
        file << "\n";
    }
    
    file.close();
    std::cout << "Trajectory saved to " << filename << std::endl;
}

int main() {
    // Create parameters
    SystemParameters params;
    AtmosphereParameters atm;
    
    // Create simulator
    Plant sim(params, &atm);
    
    // Initial state: hovering at origin with small perturbation
    State state0 = {0};
    state0[0] = 0.0;    state0[1] = 0.0;    state0[2] = -1000.0;  // Position: 1000m altitude
    state0[3] = 0.0;    state0[4] = 0.1;    state0[5] = 0.0;      // Parafoil angles
    state0[6] = 0.0;    state0[7] = 0.1;    state0[8] = 0.0;      // Cradle angles
    state0[9] = 10.0;   state0[10] = 0.0;   state0[11] = -0.5;    // Forward velocity
    state0[12] = 0.0;   state0[13] = 0.0;   state0[14] = 0.0;     // Parafoil rates
    state0[15] = 0.0;   state0[16] = 0.0;   state0[17] = 0.0;     // Cradle rates
    
    // Run simulation
    std::cout << "Running parafoil-payload simulation..." << std::endl;
    double t_final = 1000.0;  // seconds
    double dt = 0.1;          // seconds
    
    // Create control function
    ControlFunc control = testController;
    
    auto [times, states] = sim.run_simulation(state0, t_final, dt, control);
    
    std::cout << "Simulation complete: " << times.size() << " time steps" << std::endl;
    
    const State& final_state = states.back();
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Final position: x=" << final_state[0] 
              << ", y=" << final_state[1] 
              << ", z=" << final_state[2] << std::endl;
    std::cout << "Final velocity: u=" << final_state[9] 
              << ", v=" << final_state[10] 
              << ", w=" << final_state[11] << std::endl;
    
    // Save trajectory
    saveTrajectory(times, states);
    
    return 0;
}